File: impl/go/internal/engine/effects.go
package engine

import "github.com/zaibatsu/zaibatsu-go/internal/domain"

type BlockEffectTrigger string

const (
	BlockEffectInCybernet   BlockEffectTrigger = "inCybernet"
	BlockEffectUnderControl BlockEffectTrigger = "underControl"
)

func skullsFromBlockEffect(effect *domain.BlockEffect) int {
	if effect != nil && effect.Amount > 0 {
		return effect.Amount
	}
	return 1
}

func amountFromCardEffect(effect domain.CardEffect) int {
        if effect.Amount > 0 {
                return effect.Amount
        }
        return 1
}

type EffectResolution struct {
        Kind     string                `json:"kind"`
        Result    *DeleteAreaResult     `json:"result,omitempty"`
        Pawn     *domain.PawnOnBoard   `json:"pawn,omitempty"`
        PlayerID string                `json:"playerId,omitempty"`
        Amount   int                   `json:"amount,omitempty"`
}

func placePawnEffect(s *domain.GameState, gd *domain.GameData, playerID string, coord domain.Coord, pawnID string) *EffectResolution {
        if pawnID == "" || s.PlayerByID(playerID) == nil {
                return nil
        }
        if _, ok := gd.PawnByID(pawnID); !ok || s.Cybernet.PawnByID(pawnID) != nil {
                return nil
        }
        spaceID := firstOpenSpace(gd, s.Cybernet, coord, pawnID)
        if spaceID == "" {
                return nil
        }
        for i, eliminatedID := range s.Eliminated {
                if eliminatedID == pawnID {
                        s.Eliminated = append(s.Eliminated[:i], s.Eliminated[i+1:]...)
                        break
                }
        }
        pawn := &domain.PawnOnBoard{PawnID: pawnID, OwnerID: playerID, Coord: coord, SpaceID: spaceID}
        s.Cybernet.PlacePawn(pawn)
        return &EffectResolution{Kind: "place-pawn", Pawn: pawn}
}

func applyTypedBlockEffect(s *domain.GameState, gd *domain.GameData, coord domain.Coord, effect *domain.BlockEffect) *EffectResolution {
	if effect == nil {
		return nil
	}
	switch effect.Kind {
	case "area-attack":
		result := resolveDeleteAreaAtCoord(s, gd, coord, skullsFromBlockEffect(effect))
                return &EffectResolution{Kind: "area-attack", Result: &result}
        case "place-pawn":
                placed := s.Cybernet.At(coord)
                if placed == nil || placed.OwnerID == "" {
                        return nil
                }
                return placePawnEffect(s, gd, placed.OwnerID, coord, effect.Target)
	default:
		return nil
	}
}

func ApplyBlockEffect(s *domain.GameState, gd *domain.GameData, coord domain.Coord, effect *domain.BlockEffect) *EffectResolution {
	if effect == nil || effect.Kind == "" {
		return nil
	}
	return applyTypedBlockEffect(s, gd, coord, effect)
}

func ApplyBlockEffectForTrigger(s *domain.GameState, gd *domain.GameData, coord domain.Coord, trigger BlockEffectTrigger) *EffectResolution {
	placed := s.Cybernet.At(coord)
	if placed == nil {
		return nil
	}
	block, ok := gd.BlockByID(placed.BlockID)
	if !ok {
		return nil
	}
	switch trigger {
	case BlockEffectInCybernet:
		return ApplyBlockEffect(s, gd, coord, block.Effects.InCybernet)
	case BlockEffectUnderControl:
		return ApplyBlockEffect(s, gd, coord, block.Effects.UnderControl)
	default:
		return nil
	}
}

func cardEffectsForTrigger(gd *domain.GameData, cardID, trigger string) []domain.CardEffect {
        card := cardByID(gd, cardID)
        if card == nil {
                return nil
        }
        out := []domain.CardEffect{}
        for _, effect := range card.Effects {
                effectTrigger := effect.Trigger
                if effectTrigger == "" {
                        effectTrigger = "on-play"
                }
                if effectTrigger == trigger {
                        out = append(out, effect)
                }
        }
        return out
}

func applyCardEffect(s *domain.GameState, gd *domain.GameData, playerID string, coord *domain.Coord, effect domain.CardEffect) *EffectResolution {
        switch effect.Kind {
        case "place-pawn":
                if coord == nil {
                        return nil
                }
                return placePawnEffect(s, gd, playerID, *coord, effect.Target)
        case "gain-bonus":
                player := s.PlayerByID(playerID)
                if player == nil {
                        return nil
                }
                amount := amountFromCardEffect(effect)
                player.BonusCounters += amount
                return &EffectResolution{Kind: "gain-bonus", PlayerID: playerID, Amount: amount}
        default:
                return nil
        }
}

func ApplyCardEffectsForTrigger(s *domain.GameState, gd *domain.GameData, playerID, cardID, trigger string, coord *domain.Coord) []*EffectResolution {
        effects := cardEffectsForTrigger(gd, cardID, trigger)
        out := make([]*EffectResolution, 0, len(effects))
        for _, effect := range effects {
                if resolution := applyCardEffect(s, gd, playerID, coord, effect); resolution != nil {
                        out = append(out, resolution)
                }
        }
        return out
}

File: impl/ts/src/engine/effects.ts
import type { Coord } from "../domain/hex.ts";
import type { PawnOnBoard } from "../domain/pawn_board.ts";
import { blockSpace } from "../domain/pawn_board.ts";
import {
  blockById,
  pawnById,
  playerById,
  type BlockEffect,
  type CardEffect,
  type CardEffectTrigger,
  type GameData,
  type GameState,
} from "../domain/types.ts";
import { resolveDeleteAreaAtCoord, type DeleteAreaResult } from "./combat.ts";
import { canEndOn } from "./movement.ts";

export type BlockEffectTrigger = "inCybernet" | "underControl";
export type EffectResolution =
  | { kind: "area-attack"; result: DeleteAreaResult }
  | { kind: "place-pawn"; pawn: PawnOnBoard }
  | { kind: "gain-bonus"; playerId: string; amount: number };

function skullsFromEffect(effect: Exclude<BlockEffect, string>): number {
  return Number.isInteger(effect.amount) && effect.amount! > 0 ? effect.amount! : 1;
}

function amountFromCardEffect(effect: CardEffect): number {
  return Number.isInteger(effect.amount) && effect.amount! > 0 ? effect.amount! : 1;
}

function firstOpenSpace(
  s: GameState,
  gd: GameData,
  coord: Coord,
  pawnId: string,
): string | undefined {
  const placed = s.cybernet.at(coord);
  if (!placed) return undefined;
  const block = blockById(gd, placed.blockId);
  if (!block) return undefined;
  for (const space of block.spaces ?? []) {
    if (!blockSpace(block, space.id)) continue;
    if (canEndOn(gd, s.cybernet, coord, space.id, pawnId) === undefined) return space.id;
  }
  return undefined;
}

function placePawnEffect(
  s: GameState,
  gd: GameData,
  playerId: string,
  coord: Coord,
  pawnId: string | undefined,
): EffectResolution | undefined {
  if (!pawnId || !playerById(s, playerId) || !pawnById(gd, pawnId) || s.cybernet.pawnById(pawnId)) return undefined;
  const spaceId = firstOpenSpace(s, gd, coord, pawnId);
  if (!spaceId) return undefined;
  const eliminatedIndex = s.eliminated.indexOf(pawnId);
  if (eliminatedIndex >= 0) s.eliminated.splice(eliminatedIndex, 1);
  const pawn: PawnOnBoard = { pawnId, ownerId: playerId, coord: { ...coord }, spaceId };
  s.cybernet.placePawn(pawn);
  return { kind: "place-pawn", pawn };
}

function applyTypedBlockEffect(
  s: GameState,
  gd: GameData,
  coord: Coord,
  effect: Exclude<BlockEffect, string> | undefined,
): EffectResolution | undefined {
  if (!effect) return undefined;
  switch (effect.kind) {
    case "area-attack":
      return { kind: "area-attack", result: resolveDeleteAreaAtCoord(s, gd, coord, skullsFromEffect(effect)) };
    case "place-pawn": {
      const ownerId = s.cybernet.at(coord)?.ownerId ?? "";
      return ownerId ? placePawnEffect(s, gd, ownerId, coord, effect.target) : undefined;
    }
    default:
      return undefined;
  }
}

export function applyBlockEffect(
  s: GameState,
  gd: GameData,
  coord: Coord,
  effect: BlockEffect | undefined,
): EffectResolution | undefined {
  if (!effect || typeof effect === "string") return undefined;
  return applyTypedBlockEffect(s, gd, coord, effect);
}

export function applyBlockEffectForTrigger(
  s: GameState,
  gd: GameData,
  coord: Coord,
  trigger: BlockEffectTrigger,
): EffectResolution | undefined {
  const placed = s.cybernet.at(coord);
  if (!placed) return undefined;
  const block = blockById(gd, placed.blockId);
  if (!block) return undefined;
  return applyBlockEffect(s, gd, coord, trigger === "inCybernet" ? block.effects?.inCybernet : block.effects?.underControl);
}

function cardEffectsForTrigger(
  gd: GameData,
  cardId: string,
  trigger: CardEffectTrigger,
): CardEffect[] {
  const card = gd.cards.find((entry) => entry.id === cardId);
  return (card?.effects ?? []).filter((effect) => (effect.trigger ?? "on-play") === trigger);
}

function applyCardEffect(
  s: GameState,
  gd: GameData,
  playerId: string,
  coord: Coord | undefined,
  effect: CardEffect,
): EffectResolution | undefined {
  switch (effect.kind) {
    case "place-pawn":
      return coord ? placePawnEffect(s, gd, playerId, coord, effect.target) : undefined;
    case "gain-bonus": {
      const player = playerById(s, playerId);
      if (!player) return undefined;
      const amount = amountFromCardEffect(effect);
      player.bonusCounters += amount;
      return { kind: "gain-bonus", playerId, amount };
    }
    default:
      return undefined;
  }
}

export function applyCardEffectsForTrigger(
  s: GameState,
  gd: GameData,
  playerId: string,
  cardId: string,
  trigger: CardEffectTrigger,
  coord?: Coord,
): EffectResolution[] {
  return cardEffectsForTrigger(gd, cardId, trigger)
    .map((effect) => applyCardEffect(s, gd, playerId, coord, effect))
    .filter((effect): effect is EffectResolution => !!effect);
}
